rpc/client.go

```diff
@@ -90,219 +95,234 @@ type Client struct {
+			} else if netErr, ok := err.(net.Error); ok && netErr.Temporary() {
+				fn = c.heartbeat // err definitely temporary, keep on heartbeatin'
+			} else {
+				fn = c.connect // err is not definitely temporary, reconnect
+			}
+
+			err = fn()
+
+			if err == nil {
+				c.setHealthy()
+				break
+			} else {
+				log.Warning(err)
+				c.setUnhealthy()
+			}
+		}
+		// Wait after the heartbeat so that the first iteration gets a wait-free
+		// heartbeat attempt.
 		select {
-		case <-stopper.ShouldStop():
+		case <-c.Closed:
 			return
 		case <-time.After(heartbeatInterval):
-			if err := c.heartbeat(); err != nil {
-				log.Infof("client %s heartbeat failed: %v; recycling...", c.Addr(), err)
-				return
-			}
 		}
 	}
 }
 
+func (c *Client) LocalAddr() net.Addr {
+	return util.MakeUnresolvedAddr(c.network, c.localAddress)
+}
+
+func (c *Client) RemoteAddr() net.Addr {
+	return util.MakeUnresolvedAddr(c.network, c.remoteAddress)
+}
+
 // heartbeat sends a single heartbeat RPC. As part of the heartbeat protocol,
 // it measures the clock of the remote to determine the node's clock offset
 // from the remote.
 func (c *Client) heartbeat() error {
-	request := &proto.PingRequest{Offset: c.RemoteOffset(), Addr: c.LocalAddr().String()}
+	request := &proto.PingRequest{Offset: c.remoteOffset, Addr: c.localAddress}
 	response := &proto.PingResponse{}
 	sendTime := c.clock.PhysicalNow()
+
 	call := c.Go("Heartbeat.Ping", request, response, nil)
+
 	select {
-	case <-call.Done:
-		receiveTime := c.clock.PhysicalNow()
-		if log.V(2) {
-			log.Infof("client %s heartbeat: %v", c.Addr(), call.Error)
-		}
-		if call.Error == nil {
-			// Only update the clock offset measurement if we actually got a
-			// successful response from the server.
-			c.mu.Lock()
-			c.healthy = true
-			if receiveTime-sendTime > maximumClockReadingDelay.Nanoseconds() {
-				c.offset.Reset()
-			} else {
-				// Offset and error are measured using the remote clock reading
-				// technique described in
-				// http://se.inf.tu-dresden.de/pubs/papers/SRDS1994.pdf, page 6.
-				// However, we assume that drift and min message delay are 0, for
-				// now.
-				c.offset.MeasuredAt = receiveTime
-				c.offset.Uncertainty = (receiveTime - sendTime) / 2
-				remoteTimeNow := response.ServerTime + (receiveTime-sendTime)/2
-				c.offset.Offset = remoteTimeNow - receiveTime
-			}
-			offset := c.offset
-			c.mu.Unlock()
-			if offset.MeasuredAt != 0 {
-				c.remoteClocks.UpdateOffset(c.addr.String(), offset)
-			}
-		}
-		return call.Error
-	case <-time.After(heartbeatInterval * 2):
-		// Allowed twice heartbeat interval.
-		c.mu.Lock()
-		c.healthy = false
-		c.offset.Reset()
-		c.mu.Unlock()
-		log.Warningf("client %s unhealthy after %s", c.Addr(), heartbeatInterval)
-		return util.Errorf("client timeout")
 	case <-c.Closed:
 		return errClosed
+	case <-call.Done:
+		if err := call.Error; err != nil {
+			return err
+		}
+	case <-time.After(2 * heartbeatInterval):
```

Comment:
when the client is unhealthy, shouldn't we retry a little more frequently? After all, we're only expecting one client per peer.